### PR TITLE
Subscribers

### PR DESCRIPTION
**Pull request by joecohens**
- [x] Add notifications with templates to incidents and maintenance
- [x] Translation to emails
- [x] Unsubscribe from notifications

### CODE REVIEW

**Comment by joecohens** [comment]
screen shot 2015-05-24 at 7 14 05 pm

**Comment by jbrooksuk** [comment]
Boom! @joecohens this is incredible! I'll try and get this finished today :)

**Comment by jbrooksuk** [comment]
@joecohens we have the `layouts.email` template that we could extend here ;)

**Comment by dennusb** [comment]
This is pretty awesome :) Hope it will be released soon :D

**Comment by jrgifford** [comment]
This looks amazing!

**Comment by dennusb** [comment]
I'm really curious when this is going to be merged :D I'd like to try it very very much haha!

**Comment by jbrooksuk** [comment]
Once it's ready :smile: - we're extremely proud of Cachet and don't want to rush a feature like this.

**Comment by jbrooksuk** [comment]
With that said, I promise that it won't be long.

**Comment by dennusb** [comment]
Great! We will wait ^^

**Comment by jbrooksuk** [comment]
Rebased @joecohens :)

File: app/Events/CustomerHasSubscribedEvent.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
+<?php
+
+/*
+ * This file is part of Cachet.
+ *
+ * (c) James Brooks <james@cachethq.io>
```
Comment (GrahamCampbell) [inline app/Events/CustomerHasSubscribedEvent.php]: Headers need updating.

File: app/Handlers/Events/SendSubscriberVerificationEmailHandler.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+     * @param \CachetHQ\Cachet\Events\CustomerHasSubscribedEvent $event
+     *
+     * @return void
+     */
+    public function handle(CustomerHasSubscribedEvent $event)
+    {
+        $subscriber = $event->getSubscriber();
+
+        $mail = [
+            'email'     => $subscriber->email,
```
Comment (GrahamCampbell) [inline app/Handlers/Events/SendSubscriberVerificationEmailHandler.php]: over indented

File: resources/views/layout/emails.blade.php [hunk 3]
```diff
@@ -4,7 +4,7 @@
     <meta charset="utf-8"> <!-- utf-8 works for most cases -->
     <meta name="viewport" content="width=device-width"> <!-- Forcing initial-scale shouldn't be necessary -->
     <meta http-equiv="X-UA-Compatible" content="IE=edge"> <!-- Use the latest (edge) version of IE rendering engine -->
-    <title>CachetHQ</title> <!-- the <title> tag shows on email notifications on Android 4.4. -->
+    <title>{{ Setting::get('app_name') }} Status</title> <!-- the <title> tag shows on email notifications on Android 4.4. -->
```
Comment (GrahamCampbell) [inline resources/views/layout/emails.blade.php]: Should be `appName` not `app_name` everywhere.

Comment (GrahamCampbell) [inline resources/views/layout/emails.blade.php]: NVM...

File: resources/lang/de/cachet.php [hunk 4]
```diff
@@ -47,6 +47,29 @@
...
+
+    // Subscriber
+    'subscriber' => [
+        'subscribe' => 'Subscribe to get the most recent updates.',
+        'button'    => 'Subscribe',
+        'email'     => [
+            'subscribe'   => 'Subscribe to email updates.',
+            'subscribed'  => 'You\'ve been subscribed to email notifications.',
+            'verified'    => 'Your email subscription has been confirmed. Thank you!',
+            'unsuscribe'  => 'Unsuscribe from email updates.',
```
Comment (jbrooksuk) [inline resources/lang/de/cachet.php]: Unsubscribe*

File: resources/lang/de/cachet.php [hunk 5]
```diff
@@ -47,6 +47,29 @@
...
+            'unsuscribed' => 'Your email subscription has been canceled,',
```
Comment (jbrooksuk) [inline resources/lang/de/cachet.php]: Cancelled*

Comment (jbrooksuk) [inline resources/lang/de/cachet.php]: Unsubscribed*

File: resources/lang/en/cachet.php [hunk 6]
```diff
@@ -58,6 +58,20 @@
...
 
+    // Subscriber
+    'subscriber' => [
+        'subscribe' => 'Subscribe to get the most recent updates.',
+        'button'    => 'Subscribe',
+        'email'     => [
+            'subscribe'   => 'Subscribe to email updates.',
+            'subscribed'  => 'You\'ve been subscribed to email notifications, please check your email to confirm your subscription.',
+            'verified'    => 'Your email subscription has been confirmed. Thank you!',
+            'unsuscribe'  => 'Unsuscribe from email updates.',
```
Comment (jbrooksuk) [inline resources/lang/en/cachet.php]: Same as above (and for all others)

File: resources/views/emails/incidents/maintenance-html.blade.php [hunk 7]
```diff
@@ -0,0 +1,16 @@
+@extends('layout.emails')
+
+@section('preheader')
+New maintenance has been scheduled on {{ Setting::get('app_name') }}
+@stop
+
+@section('content')
+    <p>New maintenance has been scheduled on {{ Setting::get('app_name') }}</p>
```
Comment (jbrooksuk) [inline resources/views/emails/incidents/maintenance-html.blade.php]: We need to whack this in a translation, too! :)

File: .env.example [hunk 8]
```diff
@@ -11,7 +11,7 @@ DB_PASSWORD=secret
 
 CACHE_DRIVER=file
 SESSION_DRIVER=file
-QUEUE_DRIVER=sync
+QUEUE_DRIVER=database
```
Comment (GrahamCampbell) [inline .env.example]: I think we should keep sync by default, and not provide the database queue migration files.

Comment (GrahamCampbell) [inline .env.example]: Getting people to use a queue by default adds way more overhead.

Comment (jbrooksuk) [inline .env.example]: We should advocate a queue, but advise `sync` for people on cPanel etc... Imagine having hundreds or thousands of subscribers, it'd take ages to send all of the emails.

Comment (GrahamCampbell) [inline .env.example]: I don't think we should recommend the database queue if we do get people to use a queue.

Comment (GrahamCampbell) [inline .env.example]: Or at least, not include the migration by default.

Comment (jbrooksuk) [inline .env.example]: Less overhead than Beanstalk/Redis though? Are there any technical reasons?

Comment (joecohens) [inline .env.example]: You both have good points let's just remember we are making Cachet easy to install so if we make it sync by default we need to provide tools to change it. Like a setting that when you change it creates the migrations for you and automatically migrate, or composer install redis. What do you think?

Comment (joecohens) [inline .env.example]: Also if they are using queues they need to have access to the console to create a worker.

Comment (jbrooksuk) [inline .env.example]: We could of course use the task scheduler, and just have it run as a cron job in sync mode (by default)?

Comment (GrahamCampbell) [inline .env.example]: > We could of course use the task scheduler, and just have it run as a cron job in sync mode (by default)?

That's true. We could use the database queue, then schedule the queue work command every minute.

Comment (joecohens) [inline .env.example]: Totally! Forgot about the scheduling commands :)

**Comment by joecohens** [comment]
@jbrooksuk Done. Happy birthday buddy :)

@cachethq/owners PD: We still have to figure out the worker discussion, we should talk about it tomorrow. Let me know what you think :)

**Comment by jbrooksuk** [comment]
:D

**Comment by jbrooksuk** [comment]
Needs rebasing @joecohens so that the tests will pass.

**Comment by joecohens** [comment]
@jbrooksuk Rebased :)

**Comment by jbrooksuk** [comment]
This is everything, right? :)

**Comment by joecohens** [comment]
We still missing the discussion on how to handle by default the Jobs, and the scheduler

**Comment by jbrooksuk** [comment]
To Slack!

File: app/Console/Kernel.php [hunk 9]
```diff
@@ -32,6 +32,6 @@ class Kernel extends ConsoleKernel
      */
     protected function schedule(Schedule $schedule)
     {
-        //
+        $schedule->command('queue:work --sleep=3 --tries=3')->everyMinute();
```
Comment (joecohens) [inline app/Console/Kernel.php]: @jbrooksuk Done ;)

Comment (GrahamCampbell) [inline app/Console/Kernel.php]: And what happens if people aren't using the sync queue?

Comment (GrahamCampbell) [inline app/Console/Kernel.php]: Also, that doesn't look correct?

Comment (GrahamCampbell) [inline app/Console/Kernel.php]: Why would we want to sleep? Surely we want to process a single job then exit?

Comment (GrahamCampbell) [inline app/Console/Kernel.php]: Otherwise we'll end up with a gazzilian job workers.

Comment (joecohens) [inline app/Console/Kernel.php]: Yeah you right that's when running on daemon mode. I'll remove the sleep.

Comment (joecohens) [inline app/Console/Kernel.php]: Also if people aren't using the database queue they shouldn't register the schedule:run command.

Comment (joecohens) [inline app/Console/Kernel.php]: @GrahamCampbell Do you have an alternative to this?

Comment (GrahamCampbell) [inline app/Console/Kernel.php]: I'd say we should just leave people with nothing forced out of the box like laravel does. People can setup something like this if they want it imo.

Comment (jbrooksuk) [inline app/Console/Kernel.php]: I dunno. I feel like we should be able to tell them to add `schedule:run` to a cron and it's done. If they don't want it, they can use `sync` and it'll run *inline*. Easier that way?

Comment (GrahamCampbell) [inline app/Console/Kernel.php]: Just a pain if we want to start using the scheduler for anything else in the future.

Comment (jbrooksuk) [inline app/Console/Kernel.php]: How so? I thought it was supposed to make it easier?

Comment (GrahamCampbell) [inline app/Console/Kernel.php]: It makes it easier until we want to use the scheduler for something else, then we have ourselves a mess.

Comment (joecohens) [inline app/Console/Kernel.php]: So in conclusion we should remove the config or no? lol

Comment (GrahamCampbell) [inline app/Console/Kernel.php]: I think we should, James thinks we shouldn't.

In balance, we should probably keep it. I'm just warning that we may run into issues in the future.

Comment (jbrooksuk) [inline app/Console/Kernel.php]: I just want to make it easiest for the user - but if it's a technical issue, then I'm all for not doing it too :)

**Comment by GrahamCampbell** [comment]
Mergy mergy?

**Comment by jbrooksuk** [comment]
Are you ready? Are you ready for merge? Yes you are, oh yes! :notes:

**Comment by jbrooksuk** [comment]
SUBSCRIBERS! :tada:

**Comment by GrahamCampbell** [comment]
:train2:

**Comment by jrgifford** [comment]
:grin:

**Comment by joecohens** [comment]
:boom:

**Pull request merged**